### ISSUE TITLE
Jenkins build stuck in ContainerCreating

### ISSUE THREAD

**Issue by iscra** [fabric8io/fabric8#6264]
Using current vanilla Fabric8 on a fresh Kubernetes cluster (3 nodes).
Created sample Microservice project with Canary pipeline.
Fits build succeeds.
Second build always fails. Jenkins creates new pod which gets stuck in ContainerCreating, because mounting volumes fails.
Describe pod output from one of the failed builds:
```
kubectl describe pod 218d286ba5f0b
Name:		218d286ba5f0b
Namespace:	default
Node:		10.20.96.192/10.20.96.192
Start Time:	Thu, 18 Aug 2016 10:34:20 +0200
Labels:		jenkins=slave
		name=jenkins-slave-default
Status:		Pending
IP:		
Controllers:	<none>
Containers:
  slave:
    Container ID:	
    Image:		fabric8/jenkins-jnlp-client:2.2.218
    Image ID:		
    Port:		
    Args:
      97e199bb08a711f311199bd874239115ae9b7f0072fe85e1482f67504d88cc3b
      218d286ba5f0b
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Environment Variables:
      JENKINS_SECRET:			97e199bb08a711f311199bd874239115ae9b7f0072fe85e1482f67504d88cc3b
      JENKINS_LOCATION_URL:		
      JENKINS_URL:			http://jenkins:80
      JENKINS_TUNNEL:			jenkins-jnlp:50000
      JENKINS_JNLP_URL:			http://jenkins:80/computer/218d286ba5f0b/slave-agent.jnlp
      DOMAIN:				${DOMAIN}
      JENKINS_WORKFLOW_GIT_REPOS:	${JENKINS_WORKFLOW_GIT_REPOS}
      JENKINS_WORKFLOW_GIT_REPOSITORY:	https://github.com/fabric8io/jenkins-pipeline-library.git
      SONATYPE_PASSWORD:		${SONATYPE_PASSWORD}
      GITHUB_TOKEN:			${GITHUB_TOKEN}
      DOCKER_REGISTRY_USERNAME:		${DOCKER_REGISTRY_USERNAME}
      JENKINS_GOGS_EMAIL:		gogsadmin@fabric8.local
      DOCKER_REGISTRY_SERVER_ID:	${DOCKER_REGISTRY_SERVER_ID}
      DOCKER_REGISTRY_PASSWORD:		${DOCKER_REGISTRY_PASSWORD}
      DOCKER_REGISTRY_PREFIX:		${DOCKER_REGISTRY_PREFIX}
      JENKINS_JOBS_GIT_REPOSITORY:	${JENKINS_JOBS_GIT_REPOSITORY}
      SONATYPE_USERNAME:		${SONATYPE_USERNAME}
      JENKINS_GOGS_USER:		gogsadmin
      JENKINS_GOGS_PASSWORD:		RedHat$1
      GPG_PASSPHRASE:			${GPG_PASSPHRASE}
      GITHUB_AUTH_TOKEN:		${GITHUB_AUTH_TOKEN}
      NEXUS_USERNAME:			${NEXUS_USERNAME}
      NEXUS_PASSWORD:			${NEXUS_PASSWORD}
      KUBERNETES_TRUST_CERTIFICATES:	true
      KUBERNETES_NAMESPACE:		default
      DOCKER_HOST:			unix:/var/run/docker.sock
      DOCKER_AUTH_DOCKERCFG_PATH:	/home/jenkins/.docker/config.json
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  volume-0:
    Type:	HostPath (bare host directory volume)
    Path:	/var/run/docker.sock
  volume-1:
    Type:	HostPath (bare host directory volume)
    Path:	/home/jenkins/workspace
  volume-2:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-git-ssh
  volume-3:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-release-gpg
  volume-4:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-master-ssh
  volume-5:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-docker-cfg
  volume-6:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-maven-settings
  volume-7:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-hub-api-token
  jenkins-token-y66a6:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-token-y66a6
QoS Tier:	BestEffort
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  6m		6m		1	{default-scheduler }			Normal		Scheduled	Successfully assigned 218d286ba5f0b to 10.20.96.192
  4m		2m		2	{kubelet 10.20.96.192}			Warning		FailedMount	Unable to mount volumes for pod "218d286ba5f0b_default(8ed95595-651e-11e6-b32c-52540000003a)": timeout expired waiting for volumes to attach/mount for pod "218d286ba5f0b"/"default". list of unattached/unmounted volumes=[volume-5 volume-6]
  4m		2m		2	{kubelet 10.20.96.192}			Warning		FailedSync	Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "218d286ba5f0b"/"default". list of unattached/unmounted volumes=[volume-5 volume-6]

```

**Comment by rawlingsj** [comment]
@iscra I've seen this before.  I'm not sure what the cause is yet, my guess at the time was there was an issue with the cloud provider mounting volumes as I've not seen it outside of GKE but perhaps that's not the case.  

We don't actually need most of those mounts by default anymore so we can try removing them.  In the Jenkins manage config page http://jenkins.foo.io/configure , towards the bottom there's a section that has the volumes defined.  You could try deleting
- jenkins-git-ssh
- jenkins-release-gpg
- jenkins-master-ssh
- jenkins-docker-cfg
- jenkins-hub-api-token
- jenkins-maven-settings # not sure if this is needed here anymore, could try without as we add it in the jenkinsfile build pod volume mount now

**Comment by iscra** [comment]
@rawlingsj thank you for fast response!
This has actually helped a bit. I have removed the volumes from the jenkins config and only after I removed also jenkins-maven-settings, then the pod was created succesfully.

Now, it seems this build then tries to create another "buildpod" and fails again when mounting volumes. Is this some regular issue with mounting volumes and can I also fix this somewhere?

Describe log with errors:
```
Name:		buildpod-2b08159e-765a-4689-bc78-5ec590f73b1b
Namespace:	default
Node:		10.20.96.192/10.20.96.192
Start Time:	Thu, 18 Aug 2016 11:23:25 +0200
Labels:		owner=jenkins
Status:		Pending
IP:		
Controllers:	<none>
Containers:
  podstep:
    Container ID:	
    Image:		fabric8/maven-builder
    Image ID:		
    Port:		
    Command:
      /bin/sh
      -c
    Args:
      cat
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Environment Variables:
      BUILD_DISPLAY_NAME:	#2
      BUILD_ID:			2
      BUILD_NUMBER:		2
      BUILD_TAG:		jenkins-hello-2
      CLASSPATH:		
      EXECUTOR_NUMBER:		0
      HUDSON_HOME:		/var/jenkins_home
      HUDSON_SERVER_COOKIE:	dd78f13145a27081
      HUDSON_URL:		http://10.20.96.191:30098/
      JENKINS_HOME:		/var/jenkins_home
      JENKINS_SERVER_COOKIE:	dd78f13145a27081
      JENKINS_URL:		http://10.20.96.191:30098/
      JOB_NAME:			hello
      NODE_NAME:		21b7f1032d550
      DOCKER_CONFIG:		/home/jenkins/.docker/
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  volume-1:
    Type:	HostPath (bare host directory volume)
    Path:	/home/jenkins/workspace/workspace/hello
  volume-2:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-maven-settings
  volume-3:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-docker-cfg
  volume-4:
    Type:	HostPath (bare host directory volume)
    Path:	/var/run/docker.sock
  volume-5:
    Type:	HostPath (bare host directory volume)
    Path:	/root/.mvnrepository
  jenkins-token-y66a6:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	jenkins-token-y66a6
QoS Tier:	BestEffort
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  6m		6m		1	{default-scheduler }			Normal		Scheduled	Successfully assigned buildpod-2b08159e-765a-4689-bc78-5ec590f73b1b to 10.20.96.192
  4m		7s		3	{kubelet 10.20.96.192}			Warning		FailedMount	Unable to mount volumes for pod "buildpod-2b08159e-765a-4689-bc78-5ec590f73b1b_default(6a4566fb-6525-11e6-b32c-52540000003a)": timeout expired waiting for volumes to attach/mount for pod "buildpod-2b08159e-765a-4689-bc78-5ec590f73b1b"/"default". list of unattached/unmounted volumes=[volume-2 volume-3]
  4m		7s		3	{kubelet 10.20.96.192}			Warning		FailedSync	Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "buildpod-2b08159e-765a-4689-bc78-5ec590f73b1b"/"default". list of unattached/unmounted volumes=[volume-2 volume-3]

```

**Comment by rawlingsj** [comment]
dam!  so I wonder if it's just the maven settings that it's having problems mounting?  To test this we could remove the line from your jenkinsfile if you selected to copy it into your workspace when creating the app via the UI.

So in the fabric8 console left hand menu when looking at your app select the Source tab and remove this line from the Jenkinsfile https://github.com/fabric8io/jenkins-pipeline-library/blob/master/workflows/maven/CanaryReleaseAndStage/Jenkinsfile#L39

The build should trigger but if it starts the build pod we know its the maven settings volume specifically.  I'd expect the build to fail mind as it wont have the maven settings needed to deploy artifacts to nexus.

**Comment by iscra** [comment]
So, I more more in depth testing.
Both mounts fail: jenkins-docker-cfg and jenkins-maven-settings

The buildpod gets created when both are removed, but then fails in deployment:
```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.8.2:deploy (default-cli) on project service1: Deployment failed: repository element was not specified in the POM inside distributionManagement element or in -DaltDeploymentRepository=id::layout::url parameter -> [Help 1]
```

Any idea how can I track what is wrong with those mounts?

**Comment by rawlingsj** [comment]
Hmm - I'm not sure yet but lets see if we can work through it.  Are you on GKE / AWS / private cloud?  Also if you want to jump on irc.freenode.net #fabric8 it might make things a little easier while we investigate what's going on.

**Comment by jstrachan** [comment]
@iscra I hit this issue yesterday on minikube. It seems upgrading to 1.3.5 fixed it; there's been a number of volume related fixes in kubernetes around 1.3.3-1.35. Which kubernetes version are you using?

**Comment by iscra** [comment]
Yes, we suspect it too. We have a Kubernetes 1.3.3 cluster built on Core OS.
We will try to update it first to 1.3.5, do clean reinstall of fabric8 and test again.

**Comment by iscra** [comment]
I can confirm that upgrading the cluster to Kubernetes 1.3.5 has solved the issue (1.3.4 did not work).

**Comment by rawlingsj** [comment]
That's good news!  Thanks for reporting back.
